x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/stats/PlanningMetrics.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.stats;
+
+import org.elasticsearch.xpack.esql.expression.function.UnresolvedFunction;
+import org.elasticsearch.xpack.esql.plan.logical.LogicalPlan;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * This class is responsible for collecting metrics related to ES|QL planning.
+ */
+public class PlanningMetrics {
+    private Map<String, Integer> commands = new HashMap<>();
+    private Map<String, Integer> functions = new HashMap<>();
+
+    public void gatherPreAnalysisMetrics(LogicalPlan plan) {
+        plan.forEachDown(p -> add(commands, p.commandName()));
+        plan.forEachExpressionDown(UnresolvedFunction.class, p -> add(functions, p.name().toUpperCase(Locale.ROOT)));
```

Comment:
That was my intention (I'm doing the same for commands as well, that is slightly different from the legacy implementation if I'm not wrong), but I could also increment it only once per query, not sure which of the two is more useful. 
We could record both (usage per query and number of occurrences) actually, I don't think it will add significant overhead.

Reply:
After some off-line discussions, it seems both statistics could be useful.
I'll add both then

Reply:
Done